### PR TITLE
Add server_tokens configuration option

### LINKED ISSUE #825: Ability to mask the http response header Server attribute
Right now, all responses have the http header
server: gunicorn/19.0.0

For security reasons, I would like to be able to mask that so that people won't know to search for security vulnerabilities in gunicorn. Is there any way to mask it? via a setting perhaps?

### PR DESCRIPTION
**Pull request by mkgs**
Fixes #825

### CODE REVIEW

**Comment by mkgs** [comment]
Expanded on earlier PR from @mathiasverhoeven and @mprzytulski, w/ nginx-like --server-tokens true|false|string syntax.

File: docs/source/settings.rst [hunk 1]
```diff
@@ -1095,6 +1095,23 @@ The variables are passed to the the PasteDeploy entrypoint. Example::
 
 .. versionadded:: 19.7
 
+.. _server-tokens:
+
+server_tokens
+~~~~~~~~~~~~~
+
+* ``--server-tokens``
+* ``true``
```
Comment (javabrett) [inline docs/source/settings.rst]: Does not match Python `True` or value for other settings.

File: gunicorn/config.py [hunk 2]
```diff
@@ -221,6 +221,17 @@ def paste_global_conf(self):
 
         return global_conf
 
+    @property
+    def server_tokens(self):
+        s = self.settings['server_tokens'].get()
+        if s == 'true':
```
Comment (javabrett) [inline gunicorn/config.py]: Does not match Python `True` or value for other settings.

File: gunicorn/config.py [hunk 3]
```diff
@@ -221,6 +221,17 @@ def paste_global_conf(self):
...
+            return SERVER_SOFTWARE
+
+        if s == 'false':
```
Comment (javabrett) [inline gunicorn/config.py]: Might be nice to have the option to just simply disable the header altogether, not even leaking `gunicorn`.

File: gunicorn/http/wsgi.py [hunk 4]
```diff
@@ -300,10 +299,11 @@ def default_headers(self):
         headers = [
             "HTTP/%s.%s %s\r\n" % (self.req.version[0],
                 self.req.version[1], self.status),
-            "Server: %s\r\n" % self.version,
             "Date: %s\r\n" % util.http_date(),
             "Connection: %s\r\n" % connection
         ]
+        if self.cfg and self.cfg.server_tokens:
```
Comment (javabrett) [inline gunicorn/http/wsgi.py]: I haven't tested this, but with reference to my previous comments - if you can set an empty string then this is falsey and will be skipped.  Do we still need true/True/false/False?

File: gunicorn/http/wsgi.py [hunk 5]
```diff
@@ -300,10 +299,11 @@ def default_headers(self):
...
+            headers.append("Server: %s\r\n" % self.cfg.server_tokens)
```
Comment (javabrett) [inline gunicorn/http/wsgi.py]: Noting that the header order will probably change here, which I suppose is not significant.

File: tests/test_http.py [hunk 6]
```diff
@@ -226,3 +226,42 @@ def test_eof_reader_read_invalid_size():
...
+    c.set('server_tokens', 'server software')
+    response = Response(mocked_request, mocked_socket, c)
+    headers = response.default_headers()
+    assert get_server_header(headers) == 'server software'
+
+    # disabled server header
+    c.set('server_tokens', '')
+    response = Response(mocked_request, mocked_socket, c)
+    headers = response.default_headers()
+    assert get_server_header(headers) is None
```
Comment (javabrett) [inline tests/test_http.py]: Nice.

**Comment by mkgs** [comment]
see https://github.com/benoitc/gunicorn/pull/1617#issuecomment-369581668 for some additional context around the ```true|false|string``` options.

**Comment by benoitc** [comment]
following my suggestion and comments from @tilgovi & @jamadden I'm closing this issue favor of #2233

**Pull request closed without merging**